apps/explorer/src/ui/Search.tsx

```diff
@@ -0,0 +1,143 @@
+            </Text>
+        </Combobox.Option>
+    );
+}
+
+export function Search({
+    onChange,
+    onSelectResult,
+    placeholder,
+    name = 'search',
+    options = {},
+    isLoading = false,
+    inputValue,
+    value,
+}: SearchProps) {
+    const hasOptions = Object.entries(options).some(
+        ([k, v]) => !!v && Object.keys(v).length
+    );
+    return (
+        <Combobox
+            value={value}
+            onChange={onSelectResult}
+            name={name}
+            data-testid={`${name}-root`}
+            as="div"
+            className="relative flex w-full flex-col"
+        >
+            <Combobox.Input
+                data-testid={`${name}-input`}
+                displayValue={(value: SearchResult) => value.label}
+                className="text-white/0.4 border-1 h-[2rem] w-full rounded-md border-transparent bg-search-fill pl-2 text-xs leading-8 text-white focus:border-solid focus:border-sui"
+                onChange={onChange}
+                placeholder={placeholder}
+                autoComplete="off"
+                value={inputValue}
+            />
+            <button
+                type="button"
+                className="text-white/0.4 absolute inset-y-0 right-0 flex items-center border-none bg-transparent text-2xl focus:outline-none"
+            >
+                <Search16 className="bg-search-fill text-white opacity-40" />
+            </button>
+
+            {inputValue && (
+                <Combobox.Options
+                    className="absolute top-9 mt-1 max-h-[500px] w-[500px] list-none overflow-auto rounded-md bg-white p-3.5 shadow-md"
+                    data-testid={`${name}-results`}
+                >
+                    {isLoading ? (
+                        <div className="flex items-center justify-center">
+                            <LoadingSpinner />
+                        </div>
+                    ) : hasOptions ? (
+                        Object.entries(options).map(([key, results], idx) => {
+                            if (!results.length) return null;
+                            return (
+                                <div
+                                    className={
+                                        idx !==
+                                        Object.entries(options).length - 1
+                                            ? 'mb-4'
+                                            : ''
+                                    }
+                                    key={key}
+                                >
+                                    {!!results?.length && (
+                                        <div className="mb-2">
+                                            <Text
+                                                color="steel-dark"
+                                                variant="captionSmall/medium"
+                                            >
+                                                {key}
```

Comment:
Don't love using the object key as a label here. Seems easy to accidentally refactor vs having some explicit label in the results. I think a slightly better data structure would be something like...

```
[
  {label: 'Addresses', results: []}
  {label: 'Objects', results: []}
]
```

This also seems easier to determine if you have results or not.

Reply:
agreed